File: leet/rotate-list.ts
/**
 * https://leetcode.com/problems/rotate-list/
 */
function rotateRight(head: ListNode | null, k: number): ListNode | null {
  const length = count(head);
  k = k % length;

  for (let i = 1; i <= k; i++) {
    head = rotateOnce(head);
  }

  return head;
}

function rotateOnce(head: ListNode | null): ListNode | null {
  if (!head) return null;
  if (!head.next) return head;

  let pointer: ListNode | null = head!.next;
  let trailingPointer: ListNode | null = head;

  while (pointer!.next) {
    pointer = pointer!.next;
    trailingPointer = trailingPointer!.next;
  }

  trailingPointer!.next = null;
  pointer!.next = head;
  head = pointer;

  return head;
}

function count(list: ListNode | null): number {
  let counter = 0;
  while (list) {
    counter++;
    list = list.next;
  }
  return counter;
}

function toList(elements: number[]): ListNode | null {
  if (elements.length === 0) return null;

  let pointer: ListNode | null = null;
  for (let i = elements.length - 1; i >= 0; i--) {
    const node = new ListNode(elements[i]);
    node.next = pointer;
    pointer = node;
  }
  return pointer;
}

function toArray(list: ListNode | null): number[] {
  if (!list) return [];

  const result: number[] = [];
  while (list) {
    result.push(list.val);
    list = list.next;
  }
  return result;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export { toList, toArray, count, rotateOnce, rotateRight };

File: python/today4/rotate_list.py
# https://leetcode.com/problems/rotate-list/description/
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_nodes(elements: List[int]) -> Optional[ListNode]:
    if not elements:
        return None
    prev = None
    for el in reversed(elements):
        node = ListNode(el)
        node.next = prev
        prev = node
    return prev

def to_array(head: Optional[ListNode]) -> List[int]:
    if not head:
        return []
    result: List[int] = []
    while head:
        result.append(head.val)
        head = head.next
    return result

class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return head
        
        dummy = ListNode(next=head)
        pointer = dummy
        trail = dummy
        count = 0

        while pointer.next:
            count += 1
            if count > k:
                trail = trail.next
            pointer = pointer.next

        if k < count:
            pointer.next = dummy.next
            dummy.next = trail.next
            trail.next = None
        else:
            k = k % count
            if k > 0:
                pointer = dummy
                trail = dummy
                count = 0

                while pointer.next:
                    count += 1
                    if count > k:
                        trail = trail.next
                    pointer = pointer.next
    
                pointer.next = dummy.next
                dummy.next = trail.next
                trail.next = None

        return dummy.next